File: app/src/main/kotlin/com/stopptanz/app/playlist/PlaylistBuilder.kt
package com.stopptanz.app.playlist

import com.stopptanz.engine.Playlist
import com.stopptanz.engine.Track

data class ScannedFile(
    val uriString: String,
    val displayName: String,
    val mimeType: String?,
    val isDirectory: Boolean,
) {
    /** Track display name: [displayName] with its extension stripped, per the Track domain rule. */
    val trackName: String get() = displayName.substringBeforeLast('.', displayName)
}

object PlaylistBuilder {

    private val audioExtensions = setOf("mp3", "m4a", "flac", "wav", "ogg", "oga", "aac", "opus")

    fun isAudioFile(file: ScannedFile): Boolean {
        if (file.isDirectory) return false
        val mimeType = file.mimeType
        if (mimeType != null) return mimeType.startsWith("audio/")
        return file.displayName.substringAfterLast('.', "").lowercase() in audioExtensions
    }

    fun build(files: List<ScannedFile>): Playlist? {
        val tracks = files.filter(::isAudioFile).map { Track(uri = it.uriString, name = it.trackName) }
        return if (tracks.isEmpty()) null else Playlist(tracks = tracks)
    }
}

File: web/src/playlist/playlistBuilder.ts
import type { Playlist } from '../engine/playlist'

const audioExtensions = new Set(['mp3', 'm4a', 'flac', 'wav', 'ogg', 'oga', 'aac', 'opus'])

interface PickedFile {
  name: string
  type: string
}

export function trackNameFromFileName(fileName: string): string {
  const dot = fileName.lastIndexOf('.')
  return dot === -1 ? fileName : fileName.slice(0, dot)
}

export function isAudioFile(file: PickedFile): boolean {
  if (file.type) return file.type.startsWith('audio/')
  const dot = file.name.lastIndexOf('.')
  const extension = dot === -1 ? '' : file.name.slice(dot + 1).toLowerCase()
  return audioExtensions.has(extension)
}

export function buildPlaylist<F extends PickedFile>(files: F[], uriFor: (file: F) => string): Playlist | null {
  const tracks = files
    .filter(isAudioFile)
    .map((file) => ({ uri: uriFor(file), name: trackNameFromFileName(file.name) }))
  return tracks.length === 0 ? null : { tracks, shuffle: false, loop: false }
}
